sdk/extensions/Microsoft.Extensions.Azure/src/Internal/ClientRegistration.cs

```diff
@@ -71,63 +63,87 @@ public TClient GetClient(IServiceProvider serviceProvider, object options, Token
+            }
 
-                    lock (_cacheLock)
-                    {
-                       if (!_clientInitialized)
-                       {
-                           return;
-                       }
+            cachedClient = default;
+            return false;
+        }
 
-                       disposableClient = (IAsyncDisposable)_cachedClient;
+        public async ValueTask DisposeAsync()
+        {
+            if (!_clientInitialized)
+            {
+                return;
+            }
 
-                       _cachedClient = default;
-                       _clientInitialized = false;
-                    }
+            if (_cachedClient is null)
+            {
+                return;
+            }
 
-                    await disposableClient.DisposeAsync().ConfigureAwait(false);
-                }
-                else if (_disposable)
-                {
-                    Dispose();
-                }
+            IDisposable disposableClient;
+            IAsyncDisposable asyncDisposableClient;
+            lock (_cacheLock)
+            {
+                disposableClient = _cachedClient as IDisposable;
+                asyncDisposableClient = _cachedClient as IAsyncDisposable;
+                _cachedClient = default;
+            }
+
+            if (asyncDisposableClient is not null)
+            {
+                await asyncDisposableClient.DisposeAsync().ConfigureAwait(false);
+            }
+            else if (disposableClient is not null)
+            {
+                disposableClient.Dispose();
             }
         }
 
         public void Dispose()
         {
-            if (_clientInitialized)
+            if (!_clientInitialized)
             {
-                if (_disposable)
-                {
-                    IDisposable disposableClient;
-
-                    lock (_cacheLock)
-                    {
-                       if (!_clientInitialized)
-                       {
-                           return;
-                       }
+                return;
+            }
 
-                       disposableClient = (IDisposable)_cachedClient;
+            if (_cachedClient is null)
+            {
+                return;
+            }
 
-                       _cachedClient = default;
-                       _clientInitialized = false;
-                    }
+            IDisposable disposableClient;
+            IAsyncDisposable asyncDisposableClient;
+            lock (_cacheLock)
+            {
+                disposableClient = _cachedClient as IDisposable;
+                asyncDisposableClient = _cachedClient as IAsyncDisposable;
+                _cachedClient = default;
+            }
 
-                    disposableClient.Dispose();
-                }
-                else if (_asyncDisposable)
-                {
-                    DisposeAsync().GetAwaiter().GetResult();
-                }
+            if (disposableClient is not null)
```

Comment:
Should asyncDisposableClient take precedence?

Reply:
Ah I guess it makes sense since this is the Dispose call.

Reply:
Yeah, the original implementation intentionally preferred the sync version for `Dispose` and the async version for `DisposeAsync` and fell back only when there's no corresponding version.